### PR TITLE
design: document cgroups v1 and v2

### LINKED ISSUE #603: document cgroups v1 and v2
document what changes are needed to support cgroups v1 and v2

### PR DESCRIPTION
**Pull request by devimc**
document what cgroups are supported and what changes are needed in the
configuration file to support them.

fixes #603

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: design/host-cgroups.md [hunk 1]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
 [cgroupspath]: https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#cgroups-path
 
-## Summary
+# Supported cgroups
 
-| cgroup option | default? | status | pros | cons
-|-|-|-|-|-|
-| `SandboxCgroupOnly=false` | yes | legacy | Easiest to make Kata work | Unaccounted for memory and resource utilization
-| `SandboxCgroupOnly=true` | no | recommended | Complete tracking of Kata memory and CPU utilization. In Kubernetes, the Kubelet can fully constrain Kata via the pod cgroup | Requires upper layer orchestrator which sizes sandbox cgroup appropriately |
+Kata containers support cgroups `v1` and `v2`. In the following sections each cgroup is
```
Comment (GabyCT) [inline design/host-cgroups.md]: /support/supports

File: design/host-cgroups.md [hunk 2]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+
+A whole process can join a cgroup by writing its process id (`pid`) to `cgroup.procs` file,
+or join a cgroup partially by writing the task (thread) id (`tid`) to `tasks` file.
+
+Kata containers support `v1` by default and no change in the configuration file is needed.
+To know more about `cgroups v1`, see [cgroupsv1(7)][2]
+
+## Cgroups V2
+
+`Cgroups v2` are also known as unified cgroups, Unlike `cgroups v1`, the cgroups are
```
Comment (GabyCT) [inline design/host-cgroups.md]: /Unlinke/unlike

File: design/host-cgroups.md [hunk 3]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
-|-|-|-|-|-|
-| `SandboxCgroupOnly=false` | yes | legacy | Easiest to make Kata work | Unaccounted for memory and resource utilization
-| `SandboxCgroupOnly=true` | no | recommended | Complete tracking of Kata memory and CPU utilization. In Kubernetes, the Kubelet can fully constrain Kata via the pod cgroup | Requires upper layer orchestrator which sizes sandbox cgroup appropriately |
+Kata containers support cgroups `v1` and `v2`. In the following sections each cgroup is
+described briefly and what changes are needed in kata containers to support it.
+
+## Cgroups V1
+
+`Cgroups V1` are under a [tmpfs][1] filesystem mounted at `/sys/fs/cgroup`, where each cgroup is
+mounted under a separated cgroup filesystem. `Cgroups v1` hierarchy may look like the following
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: s/separated/separate/

File: design/host-cgroups.md [hunk 4]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+│   └── tasks
+├── pids
+│   ├── cgroup.procs
+│   └── tasks
+└── systemd
+    ├── cgroup.procs
+    └── tasks
+```
+
+A whole process can join a cgroup by writing its process id (`pid`) to `cgroup.procs` file,
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: s/whole//

File: design/host-cgroups.md [hunk 5]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+or join a cgroup partially by writing the task (thread) id (`tid`) to `tasks` file.
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: s/`tasks` file/the `tasks` file/

File: design/host-cgroups.md [hunk 6]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+├── memory.stat
+├── memory.swap.current
+├── memory.swap.events
+├── memory.swap.max
+├── pids.current
+├── pids.events
+└── pids.max
+```
+
+Same as `cgroups v1`, a whole process can join the cgroup by writing its process id (`pid`) to
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: s/whole//

File: design/host-cgroups.md [hunk 7]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+`cgroup.procs` file, or join a cgroup partially by writing the task (thread) id (`tid`) to `cgroup.threads` file.
+
+Kata containers don't support `v2` by default, `sandbox_cgroup_only` must be set to `true` in
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: s/don't/does not/ (or doesn't).

File: design/host-cgroups.md [hunk 8]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+the configuration file.
+To know more about `cgroups v2`, see [cgroupsv2(7)][3]
+
+### Limitations
+
+Officially `cgroups v2` are not yet supported in the guest, but they can work with a fedora 31 rootfs
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: Maybe rephrase to something like:

> Many Linux distributions do not yet support `cgroups v2`, as it is quite a recent addition...

File: design/host-cgroups.md [hunk 9]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
+image and the latest kata agent (at the time this document was written the latest version
+was `1.11.0-alpha0`). For more information about the status of this feature see [issue #2494][4]
+
+# Summary
+
+| cgroup option | default? | status | pros | cons | cgroups
+|-|-|-|-|-|-|
+| `SandboxCgroupOnly=false` | yes | legacy | Easiest to make Kata work | Unaccounted for memory and resource utilization | v1
+| `SandboxCgroupOnly=true` | no | recommended | Complete tracking of Kata memory and CPU utilization. In Kubernetes, the Kubelet can fully constrain Kata via the pod cgroup | Requires upper layer orchestrator which sizes sandbox cgroup appropriately | v1, v2
```
Comment (grahamwhaley) [inline design/host-cgroups.md]: I'm struggling with the status of `recommended`, when it is not the default today. Maybe we can change that to `supported`, and if we have a timeline for when we might make it the default, add that in as well?

File: design/host-cgroups.md (hunk 1 above)
Comment (jodh-intel) [inline design/host-cgroups.md]: s/containers/Containers/

File: design/host-cgroups.md [hunk 10]
```diff
@@ -200,9 +204,132 @@ utilized unaccounted for on the host.
...
 
-## Summary
+# Supported cgroups
 
-| cgroup option | default? | status | pros | cons
-|-|-|-|-|-|
-| `SandboxCgroupOnly=false` | yes | legacy | Easiest to make Kata work | Unaccounted for memory and resource utilization
-| `SandboxCgroupOnly=true` | no | recommended | Complete tracking of Kata memory and CPU utilization. In Kubernetes, the Kubelet can fully constrain Kata via the pod cgroup | Requires upper layer orchestrator which sizes sandbox cgroup appropriately |
+Kata containers support cgroups `v1` and `v2`. In the following sections each cgroup is
+described briefly and what changes are needed in kata containers to support it.
```
Comment (jodh-intel) [inline design/host-cgroups.md]: s/kata containers/Kata Containers/

File: design/host-cgroups.md (hunk 3 above)
Comment (jodh-intel) [reply design/host-cgroups.md]: > A `Cgroups v1` hierarchy may look like ...

File: design/host-cgroups.md (hunk 7 above)
Comment (jodh-intel) [inline design/host-cgroups.md]: I'd rephrase this slightly. Something like:

> For backwards compatibility Kata Containers defaults to supporting cgroups v1 by default. To change this to v2, set `sandbox_cgroup_only=true` in the `configuration.toml` file.

File: design/host-cgroups.md (hunk 9 above)
Comment (devimc) [reply design/host-cgroups.md]: I agree, we are waiting for the k8s podoverhead feature that may land in the next alpha release.... ?

cc @amshinde @egernst

File: design/host-cgroups.md [hunk 11]
```diff
@@ -200,9 +204,134 @@ utilized unaccounted for on the host.
...
+For backwards compatibility Kata Containers defaults to supporting cgroups v1 by default.
+To change this to `v2`, set `sandbox_cgroup_only=true` in the `configuration.toml` file.
+To know more about `cgroups v2`, see [cgroupsv2(7)][3]
+
+### Limitations
+
+Many Linux distributions do not yet support `cgroups v2`, as it is quite a recent addition,
+but they can work with a fedora 31 rootfs image and the latest `kata-agent` (at the time this
+document was written the latest version was `1.11.0-alpha0`). For more information about
+the status of this feature see [issue #2494][4]
```
Comment (jodh-intel) [inline design/host-cgroups.md]: Nit: Missing period.

File: design/host-cgroups.md [hunk 12]
```diff
@@ -200,9 +204,134 @@ utilized unaccounted for on the host.
...
+`cgroup.threads` file.
+
+For backwards compatibility Kata Containers defaults to supporting cgroups v1 by default.
+To change this to `v2`, set `sandbox_cgroup_only=true` in the `configuration.toml` file.
+To know more about `cgroups v2`, see [cgroupsv2(7)][3]
+
+### Limitations
+
+Many Linux distributions do not yet support `cgroups v2`, as it is quite a recent addition,
+but they can work with a fedora 31 rootfs image and the latest `kata-agent` (at the time this
```
Comment (jodh-intel) [inline design/host-cgroups.md]: I think we may have lost @grahamwhaley's comment about making this more generic (removing mention of Fedora 31)?

File: design/host-cgroups.md [hunk 13]
```diff
@@ -200,9 +204,134 @@ utilized unaccounted for on the host.
...
+└── systemd
+    ├── cgroup.procs
+    └── tasks
+```
+
+A process can join a cgroup by writing its process id (`pid`) to `cgroup.procs` file,
+or join a cgroup partially by writing the task (thread) id (`tid`) to the `tasks` file.
+
+Kata Containers supports `v1` by default and no change in the configuration file is needed.
+To know more about `cgroups v1`, see [cgroupsv1(7)][2]
```
Comment (jodh-intel) [inline design/host-cgroups.md]: Nit: Missing period.

File: design/host-cgroups.md [hunk 14]
```diff
@@ -7,7 +7,11 @@
   - [`SandboxCgroupOnly` disabled (default, legacy)](#sandboxcgrouponly-disabled-default-legacy)
     - [What does this method do?](#what-does-this-method-do)
       - [Impact](#impact)
-  - [Summary](#summary)
+- [Supported cgroups](#supported-cgroups)
+  - [Cgroups V1](#cgroups-v1)
+  - [Cgroups V2](#cgroups-v2)
+    - [Limitations](#limitations)
```
Comment (jodh-intel) [inline design/host-cgroups.md]: I wonder if "Distro Support" would make this section clearer as it isn't explaining a limitation of "cgroups v2", but more an issue with distro support of it.

**Comment by devimc** [comment]
@jodh-intel thanks, changes applied

File: design/host-cgroups.md (hunk 9 above)
Comment (egernst) [reply design/host-cgroups.md]: Correct.  The K8S PodOverhead feature [should move to beta in 1.18](https://github.com/kubernetes/kubernetes/issues/88339).  I wouldn't want to move to a default until this is in place, and until we have appropriate RuntimeClass definitions available which utilize Overhead.
